Feature: add initial GA4 support

khawkins98 opened the pull request:
This addresses https://github.com/visual-framework/vf-core/issues/1428 and supports classic GA and GA4. I've done some testing, but further testing is recommended (note I've suggested this as 1.1.0-rc.0)

khawkins98 committed:
Feature: add initial GA4 support

This addresses https://github.com/visual-framework/vf-core/issues/1428 and supports classic GA and GA4. I've done some testing, but further testing is recommended (note I've suggested this as 1.1.0-rc.0)

sandykadam commented:
Thanks @khawkins98, this is great. In-fact I was going to start discussion with team to prioritise this as G4 rolling out next year.

Will update you on testing of this.

nitinja commented:
Great, accepted, thanks

sandykadam commented on components/vf-componenet-rollup/scripts.js:
```diff
@@ -20,7 +20,8 @@ vfBannerElixir();
 let vfGaTrackOptions = {
-  vfGaTrackPageLoad: true
+  vfGaTrackPageLoad: true,
+  vfGa4MeasurementId: "G-N5ZLH2KNJP"
```
Is this intentional or left by mistake.

khawkins98 replied:
This is used by the visual-framework.dev doc site, but it probably raises more chances for confusion than help. So I made a clear generic that needs to be updated.

khawkins98 committed:
Set a generic vfGa4MeasurementId

oss6 commented on components/vf-analytics-google/vf-analytics-google.js:
```diff
@@ -116,21 +122,30 @@ function vfGetMeta(metaName) {
 
+  if (typeof gtag === "undefined") {
+    // if the site is still using legacy GA, set a dummy gtag function so we don't have to add a bunch of if statements
+    var gtag = function() {};
```
I encountered an issue when testing this. When using `var` the `gtag` is hoisted, hence it'll be `undefined` before the check (see https://stackoverflow.com/a/36911203).
A solution is to use the following `window.gtag = function() {}`

khawkins98 replied:
Hi Ossama! By pure coincidence I've been working through this code this morning, I'll be revising this PR with some fixes that include this issue.

khawkins98 committed:
Merge branch 'visual-framework:develop' into feature/ga4-support

oss6 commented on components/vf-analytics-google/vf-analytics-google.js:
```diff
@@ -189,6 +209,7 @@ function vfGaInit(vfGaTrackOptions) {
   // standard google analytics bootstrap
   if (vfGaTrackOptions.vfGaTrackPageLoad) {
     ga("send", "pageview");
```
Here `ga` may not be defined if one moves to ga4. I think we may need to use the same checks for `gtag`

khawkins98 committed:
Enhance vfGaLogMessage and make better use of it

khawkins98 committed:
Align gtag to GA4 event logging

khawkins98 committed:
Make vfGaIndicateUnloaded less agressive

khawkins98 committed:
Improve gtag and ga checking

Reduce times when double checks may occur.

khawkins98 committed:
Make gtag and ga check less likely to overwrite valid functions

khawkins98 commented:
I've just pushed a fair number of changes (broken down into separate thematic commits) -- these address some issues that @oss6 spotted, in addition to making it a bit more resilient, improving the optional debug command and sending more appropriate events to GA4.

There's also a documentation task to be done, as GA4 has higher expectations on configuring the Google Analytics Profile for custom events (see section "IMPORTANT: Register custom definitions" at https://www.analyticsmania.com/post/how-to-track-custom-events-with-google-analytics-4/#custom-definitions) 

~I also want to point out in the readme that this approach does not work well at all when attempting to combine it with GA4 that is loaded through Google Tag Manager -- you should use the local JS approach:~ (I was wrong, see comment below)

khawkins98 commented:
> I also want to point out in the readme that this approach does not work well at all when attempting to combine it with GA4 that is loaded through Google Tag Manager -- you should use the local JS approach:

I misspoke about that, Google Tag Manager should work fine. In my case, there was another bug causing the two clash.

khawkins98 committed:
Move all GA4 events into "click" action

It's most appropriate and saves additional configuration in the GA4 property where custom "dimension" configuration would be required.

khawkins98 committed:
Add event_label for UA compatibility

khawkins98 commented:
I'm still working my way through some of the complication between GA4 + UA + gtag + Google Tag Manager and am posting some things here before I forget them over the long weekend here:

1. In _principle_ you can use this gtag.js method with Google Tag Manager, but it seems unreliable and often leads to race conditions where this code may fail to work -- I think it's not realistic to support using this + Google Tag Manager. BUT, this is consistent with most of my reading on best practice. It is summed up well at https://measureschool.com/google-tag-manager-vs-global-site-tag/
    - > However, both these tools utilize some common resources that can lead to conflict in some cases. For example, the data layer is one such resource that is used by both tools. Therefore, it is important to be careful while adding both of them together. You will also need to test the implementation of tools by checking whether the correct data is sent to these tools. Overall, the process is doable but complicated. That’s why I would recommend not to install them together on the same page. 
    - That is: if you want to use Google Tag Manager, you should use that for your events. If you want to use gtag.js + custom JS, then vf-google-analytics is good for you. This code will provide particular value if you need a consistent way to collect event data across many websites, which is what we're after.
2. The basics of `vf-analytics-google` will be plug-and-play if using gtag+GA4, but to get the full value out if you need to add (currently) 4 "custom definitions": 
    -  ![grafik](https://user-images.githubusercontent.com/928100/193329800-4c750a70-1d77-4623-ba00-87db4d608511.png)
    - This is an important step that will need to be done on each GA4 profile (adding the Custom Definitions will not retroactively collect data).
3. Good news: None of the changes in this PR should result in regressions for sites using UA + ga.js

This all needs to be reflected in the REAMDE.md

oss6 commented:
Hi all, I tested the changes with the COVID-19 Data Portal, and everything seems to be fine :+1:

khawkins98 commented:
> Hi all, I tested the changes with the COVID-19 Data Portal, and everything seems to be fine 👍

Nice. I'm hoping to get the readme updated soon.

khawkins98 committed:
Docs and UA compatability

Adds documentation and makes log messages more consistent with existing data captured.

khawkins98 commented:
I've updated the documentation and made a few small changes to the way event messages are sent so the data format remains consistent in UA.

I believe this should now be ready for merging.

For the amount of effort that went into this PR, the diff reads quite straighforward!

khawkins98 committed:
Note that vfGa4MeasurementId is not required for most use cases

khawkins98 committed:
Further documentation tweaks
